src/metabase/search/scoring.clj

```diff
@@ -138,23 +136,44 @@
      (count query-tokens)))
 
 (defn fullness-scorer
-  "How much of the *result* is covered by the search query?"
+  "How much of the result is covered by the search query?"
   [query-tokens match-tokens]
   (let [match-token-count (count match-tokens)]
     (if (zero? match-token-count)
       0
       (/ (occurrences query-tokens match-tokens matches-in?)
          match-token-count))))
 
+(defn- prefix-counter
+  [query-string item-string]
+  (reduce
+   (fn [cnt [a b]]
+     (if (= a b) (inc cnt) (reduced cnt)))
+   0
+   (map vector query-string item-string)))
+
+(defn- count-token-chars
+  "Tokens is a seq of strings, like [\"abc\" \"def\"]"
+  [tokens]
+  (reduce
+   (fn [cnt x] (+ cnt (count x)))
+   0
+   tokens))
+
+(defn prefix-scorer
+  "How much does the search query match the beginning of the result? "
+  [query-tokens match-tokens]
+  (let [query (str/lower-case (str/join " " query-tokens))
+        match (str/lower-case (str/join " " match-tokens))]
+    (/ (prefix-counter query match)
+       (count-token-chars query-tokens))))
+
 (def ^:private match-based-scorers
-  [{:scorer consecutivity-scorer
-    :weight 1}
-   {:scorer total-occurrences-scorer
-    :weight 1}
-   {:scorer fullness-scorer
-    :weight 1/2}
-   {:scorer exact-match-scorer
-    :weight 2}])
+  [{:scorer exact-match-scorer :name "exact-match" :weight 4}
+   {:scorer consecutivity-scorer :name "consecutivity" :weight 2}
+   {:scorer total-occurrences-scorer :name "total-occurrences" :weight 2}
+   {:scorer fullness-scorer :name "fullness" :weight 1}
+   {:scorer prefix-scorer :name "prefix" :weight 1}])
 
```

Comment:
Sorted these by their weights. Also added prefix-scorer.

Reply:
Adding `name` *almost* makes me want a macro

```
(scorer consecutivity 2) ;; => {:scorer consecutivity-scorer :name "consecutivity" :weight 2}
```

but probably it's cleaner to just accept the repetition and not worry about it